src/gadgets/modexp/mod.rs

```diff
@@ -0,0 +1,225 @@
+use ethereum_types::U256;
+
+use crate::{cs::traits::cs::ConstraintSystem, field::SmallField};
+
+use super::{
+    boolean::Boolean,
+    traits::selectable::Selectable,
+    u256::UInt256,
+    u32::UInt32,
+    u512::UInt512,
+};
+
+const U256_MAX_BITS: usize = 256;
+const U512_MAX_BITS: usize = 512;
+const U256_MAX_LIMBS: usize = 8;
+const U512_MAX_LIMBS: usize = 16;
+
+const MAX_BINARY_SEARCH_ITERATIONS: usize = 33;
+
+fn convert_limb_to_u512<F, CS>(cs: &mut CS, limb: &UInt32<F>) -> UInt512<F>
+where
+    F: SmallField,
+    CS: ConstraintSystem<F>,
+{
+    let mut u512 = UInt512::zero(cs);
+    u512.inner[0] = limb.clone();
+    u512
+}
+
+fn convert_u256_to_u512<F, CS>(cs: &mut CS, u256: &UInt256<F>) -> UInt512<F>
+where
+    F: SmallField,
+    CS: ConstraintSystem<F>,
+{
+    let mut u512 = UInt512::zero(cs);
+    u512.inner[..8].copy_from_slice(&u256.inner);
+    u512
+}
+
+/// Returns `true` if `a >= b`, `false` otherwise.
+/// Here, `a` and `b` are represented as `UInt512<F>` and `UInt256<F>` respectively.
+fn u512_geq_than_u256<F, CS>(cs: &mut CS, a: &UInt512<F>, b: &UInt256<F>) -> Boolean<F>
+where
+    F: SmallField,
+    CS: ConstraintSystem<F>,
+{
+    let mut limbs = [UInt32::zero(cs); 16];
+    // Put b.inner into first 8 elements
+    limbs[..8].copy_from_slice(&b.inner);
+
+    let b_u512 = UInt512::from_limbs(limbs);
+    // If a >= b, then b-a should overflow or equal to 0
+    let (sub, overflow) = b_u512.overflowing_sub(cs, a);
+    let a_equal_b = sub.is_zero(cs);
+
+    overflow.or(cs, a_equal_b)
+}
```

Comment:
could it potentially be cheaper to do this with `a.to_low()` rather than making a `UInt512` for `b` and incurring the extra constraints during the `sub`?

Reply:
what i mean here is something like this:

```rust
let high = a.to_high();
let under_256 = high.is_zero(cs);
let over_256 = under_256.negated(cs);
let low = a.to_low();
let (sub, overflow) = b.overflowing_sub(cs, low);
let a_equal_b = sub.is_zero(cs);
Boolean::multi_or(cs, &[overflow, a_equal_b, over_256])
```

i think it may be a little cheaper

Reply:
Yeah, sub for two `UInt256<F>` should be indeed much cheaper than for two `UInt512<F>`, fixed!